[2.8][Form] Deprecate alias tag option

wouterj opened the pull request:
FQCN should be used since 2.8 instead, so a deprecation error should be triggered when the `alias` setting is used.

Furthermore, the name of the option doesn't make much sense for form types (as it's the alias of the field it applies to), so I renamed it to `extended_type`. I'm open to any other suggestions.

| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

wouterj committed:
Deprecated alias setting of form tags

stof commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/FormPass.php:
```diff
@@ -53,11 +54,15 @@ public function process(ContainerBuilder $container)
-            $alias = isset($tag[0]['alias'])
-                ? $tag[0]['alias']
-                : $serviceId;
+            $extendedType = $serviceId;
```
we need a deprecation to be triggered when the service id is used, because such case makes no sense actually.

WouterJ replied:
I agree that it doesn't mean anything, but it's the default at the moment. Do you think we should require setting the `extended_type` attribute and trigger an exception if it is not set?

Tobion replied:
@WouterJ yes if neither the alias or extended_type is set, it should trigger a deprecation as well. in 3.0 it should then throw an exception.

Tobion replied:
please move this to the `else` block which makes it more clear what code can be removed later

stof commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/FormPass.php:
```diff
@@ -53,11 +54,15 @@ public function process(ContainerBuilder $container)
-                ? $tag[0]['alias']
-                : $serviceId;
+            $extendedType = $serviceId;
+            if (isset($tag[0]['extended_type'])) {
```
I'm not a fan of this renaming:

- it does not really bring anything new
- the new name will not look good in XML configuration files, as underscores in attribute names is not common at all. The current config looks better.

WouterJ replied:
Option names are normalized: https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php#L240

I really don't like the `alias` name:

 * It's not clear that this is the alias of the form it extends
 * If that's clear, it makes no sense anymore as there is no alias anymore (it's a FQCN from now on)

stof replied:
Oh, I forgot that.

stof commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/FormPass.php:
```diff
@@ -34,18 +34,19 @@ public function process(ContainerBuilder $container)
             // in Symfony 3.0
             // Deprecation errors are triggered in the form registry
             if (isset($tag[0]['alias'])) {
+                @trigger_error(sprintf('The alias option of the form.type tag is deprecated since version 2.8 and will be removed in 3.0. Use the fully-qualified type class name "%s" instead.', $serviceDefinition->getClass()));
```
no need to trigger a deprecation warning here. It only makes it harder to support both the old and the new APIs, but having to set the tag selectively. Having an unused attribute does not hurt. And the form component will already trigger a deprecation warning in case you access the form type by its alias.

stof commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/FormPass.php:
```diff
@@ -34,18 +34,19 @@ public function process(ContainerBuilder $container)
 
         foreach ($container->findTaggedServiceIds('form.type') as $serviceId => $tag) {
+            // Support type access by FQCN
+            $serviceDefinition = $container->findDefinition($serviceId);
```
You can keep using ``getDefinition`` here. No need to use ``findDefinition`` to resolve aliases, as aliases cannot be tagged, and you are not altering the container in the loop.

WouterJ commented:
PR updated

Tobion commented:
Status: Needs Work

WouterJ commented:
Unless I missed something, I've now finished this PR:

 * Added CHANGELOG + UPGRADE notes
 * Deprecated the `alias` option for form types, but didn't trigger any deprecation notice as extra attributes don't cause problems (and it's not possible with the current BC layer to remove `alias` argument usages in the Symfony bundles)
 * Renamed `alias` to `extended_type` for form type extensions (which is consistent with the `FormTypeExtensionInterface#getExtendedType()` method btw)
 * Changed all usages of the deprecated options in the Symfony code base
 * Deprecated the fallback to service id for `alias`/`extended_type`, `extended_type` is now required

WouterJ commented:
Status: Needs Review

WouterJ commented:
Build failures are not related to this PR btw

Tobion commented on UPGRADE-2.8.md:
```diff
@@ -235,6 +235,23 @@ Form
    implementation to define the right type.
 
+ * The alias option of the `form.type_extension` tag is deprecated in favor of
+   the `extended_type` option.
```
Maybe also use `extended_type`/`extended-type` esp. since the following is code is xml which will cause confusion.

wouterj committed:
Explicitely mention extended-type as well

wouterj committed:
Minimalize changes when removing deprecated features

WouterJ commented:
Fixed, thanks @Tobion

Tobion commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/FormPass.php:
```diff
@@ -53,11 +53,18 @@ public function process(ContainerBuilder $container)
-            $alias = isset($tag[0]['alias'])
-                ? $tag[0]['alias']
-                : $serviceId;
+            $extendedType = null;
```
this is useless

Tobion replied:
fixed

WouterJ replied:
Well, it is useless, but it doesn't hurt that much and reduces the changes to be done in 3.0.

Tobion replied:
in 3.0 we will throw an exception, there is no default

Tobion commented:
Thank you @WouterJ.
